File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> m = new HashMap<>();
        int res = 0;
        
        for(int i = 0, cur = -1; i < s.length(); i++) {
            char c = s.charAt(i);
            cur = Math.max(m.getOrDefault(c, -1), cur);
            m.put(c, i);
            res = Math.max(res, i - cur);
        }

        return res;
    }
}

File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 0
        seen = {}
        l = -1
        
        for i, c in enumerate(s):
            l = max(l, seen.get(c, -1))
            res = max(res, i - l)
            seen[c] = i
                
        return res
        